### DISCUSSION TITLE
What is the plan for missing/incompatible lifecycle mappings?

### DISCUSSION THREAD

**Discussion by guw** [eclipse-m2e/m2e-core#919] [Q&A]
I am testing M2E 2.0 and I noticed that a lot of Maven projects don't build. The problem seems to be old lifecycle extensions, which were developed with semantic versioning in mind.

```
  Cannot satisfy dependency:
    From: Plexus Metadata Generation 0.15.0.201207090126 (org.sonatype.m2e.plexus.annotations 0.15.0.201207090126)
    To: osgi.bundle; org.eclipse.m2e.core [1.0.0,2.0.0)
```

```
  Cannot satisfy dependency:
    From: Tycho Project Configurators 0.9.0.201811261502 (org.sonatype.tycho.m2e 0.9.0.201811261502)
    To: osgi.bundle; org.eclipse.m2e.core [1.4.0,2.0.0)
```

Is there a plan to upgrade all these lifecycle mappings or is there a new repository/catalog to configure?

**Comment by laeubi** [comment]
In general there are the following options:

1. Ask (or even help) the specific connector project to upgrade to m2eclipse 2.x
2. If the project is abandoned, covers a common plugin (e.g. maven-* or similar widely used) and has a suitable license we can discuss to adopt the connector
3. Check if the connector is actually needed, or if something similar can be archived with a lifecycle-mapping.xml together with incremental build support in the plugin itself.

If I found out correctly your particular case is https://github.com/tesla/m2eclipse-plexus from a first look option 3 seems suitable here, but sadly there is not much description about what this actually offers.

**Reply by guw** [reply]
Is there some doc about `lifecycle-mapping.xml`?

Would it make sense go build a community managed repo of those for Maven plug-ins? I am probably going to run into a few of those and forking and releasing each one individually has a lot of overhead.

**Reply by laeubi** [reply]
> Is there some doc about `lifecycle-mapping.xml`?

You can find here how to make a maven-plugin itself compatible with m2eclipse, that way no special connector is required:
https://www.eclipse.org/m2e/documentation/m2e-making-maven-plugins-compat.html

> Would it make sense go build a community managed repo of those for Maven plug-ins? I am probably going to run into a few of those and forking and releasing each one individually has a lot of overhead.

Actually m2e-core is somewhat a "community managed repo" and help is welcome here of course! So if you cam along such connector, you can simply suggest that we adopt it, the best would be providing a PR so we can create a CQ for it. This usually takes some time, so its best to start this process early.

**Reply by HannesWell** [reply]
Option 2 is something I wanted to discuss/suggest as well. I think it would be good if M2E would provide the connectors for the most commonly used/relevant standard Maven Plug-ins, where it is necessary. Mostly from the `org.apache.maven` group, but if we are eager we could consider mojohouse as well.

Another alternative would be to make maven more cooperative with IDEs by default, as discussed here:
https://github.com/eclipse-m2e/m2e-core/discussions/876

But however it is achieved I think it would be good if the most commonly used standard Maven-plugins work out of the box without the need to install extra plug-ins. The more a plug-in is used the more effort would be justified.

**Reply by guw** [reply]
I've updated the component and published it here:
https://eclipseguru.github.io/m2eclipse-components/

How does it get added to the discovery catalog?

**Reply by laeubi** [reply]
I think @fbricon can tell best. As far as I know one can even have custom catalogs, or you can simply install it with `Help > Install New Software`

**Reply by fbricon** [reply]
https://github.com/takari/m2e-discovery-catalog/blob/master/org.eclipse.m2e.discovery.oss/src/main/resources-filtered/connectors.xml needs to be updated/curated for m2e 2.x support.

Once done, I can look into publishing a new catalog (will need to change some build stuff)

**Comment by guw** [comment]
Tycho seems to be affected as well:
```
  Cannot satisfy dependency:
    From: Tycho Project Configurators 0.9.0.201811261502 (org.sonatype.tycho.m2e 0.9.0.201811261502)
    To: osgi.bundle; org.eclipse.m2e.core [1.4.0,2.0.0)
```

**Reply by laeubi** [reply]
Tycho already ships with `lifecycle-mapping.xml` so there should no connector be required anymore.

**Reply by HannesWell** [reply]
> Tycho already ships with `lifecycle-mapping.xml` so there should no connector be required anymore.

Although Tycho ships with embedded lifecycle-mappings it can happen that the Tycho connector is suggested because it also provides a mapping for Maven's clean plug-in, but this is somewhat a false positive as explained here:
https://github.com/eclipse-m2e/m2e-core/issues/917#issuecomment-1251419467

**Reply by guw** [reply]
Is this depending on the Tycho version?

**Reply by cdietrich** [reply]
needs 2.7.4+ to my knowledge
